### PR TITLE
$now on sqlite

### PR DESCRIPTION
**Pull request by sjvans**
- [ ] changelog
- [ ] check deploy to postgres

---

given
```cds
entity Foo {
  $now as now : Timestamp
}
```
old `ISO(CURRENT_TIMESTAMP)` results in `2023-09-26T19:26:14.000Z` -> 2 hours off and no millisecond resolution
whereas new `STRFTIME('%Y-%m-%dT%H:%M:%fZ','NOW')` results in `2023-09-26T21:24:10.672Z`

alternatively, we could look into something like
```js
dbc.function('ISO', { deterministic: true }, d => {
  if (!d) return d
  if (!d.match(/Z$/)) d += 'Z'
  return d && new Date(d).toISOString()
})
```
but i believe `CURRENT_TIMESTAMP` will never have ms precision on sqlite.

where else would we need to adjust?

would it be possible to create a function `CURRENT_UTCTIMESTAMP` like we (should) use on hana?

### CODE REVIEW

File: db-service/lib/SQLService.js [hunk 1]
```diff
@@ -170,7 +170,7 @@ class SQLService extends DatabaseService {
       // REVISIT: this is a hack the target of $now might not be a timestamp or date time
       // Add input converter to CURRENT_TIMESTAMP inside views using $now
       if(/^CREATE VIEW.* CURRENT_TIMESTAMP[( ]/is.test(query)) {
-        query = query.replace(/CURRENT_TIMESTAMP/gi, 'ISO(CURRENT_TIMESTAMP)')
+        query = query.replace(/CURRENT_TIMESTAMP/gi, "STRFTIME('%Y-%m-%dT%H:%M:%fZ','NOW')")
```
Comment (johannes-vogel) [inline db-service/lib/SQLService.js:173]: What's the difference between `CURRENT_TIMESTAMP` and `NOW`?

The `ISO` function is a js plugin and simply creates a `new Date().toISOString()`

Comment (sjvans) [reply db-service/lib/SQLService.js:173]: `CURRENT_TIMESTAMP` only has s precision, afaik

Comment (sjvans) [reply db-service/lib/SQLService.js:173]: > The `ISO` function is a js plugin and simply creates a `new Date().toISOString()`

yes, see addition to description for alternative solution by modifying the `ISO` function.

File: db-service/lib/SQLService.js [hunk 2]
```diff
@@ -170,7 +170,7 @@ class SQLService extends DatabaseService {
       // REVISIT: this is a hack the target of $now might not be a timestamp or date time
       // Add input converter to CURRENT_TIMESTAMP inside views using $now
       if(/^CREATE VIEW.* CURRENT_TIMESTAMP[( ]/is.test(query)) {
```
Comment (sjvans) [inline db-service/lib/SQLService.js:172]:
```suggestion
      if(this.options.kind === 'sqlite' && /^CREATE VIEW.* CURRENT_TIMESTAMP[( ]/is.test(query)) {
```
i need this locally to not apply the patch during deployment to postgres... why hasn't anybody else run into this?

Comment (johannes-vogel) [reply db-service/lib/SQLService.js:172]: I guess we don't test the replacement yet.

**Comment by sjvans** [comment]
@danjoa @BobdenOs @patricebender any opinions?

**Comment by patricebender** [comment]
> @danjoa @BobdenOs @patricebender any opinions?

If the requirement is to get to `ms` precision, this is definitely an improvement. Before this change, it looks like we used `UTC` time. Do we want to present the date as local time in this case?

**Comment by sjvans** [comment]
> > @danjoa @BobdenOs @patricebender any opinions?
> 
> If the requirement is to get to `ms` precision, this is definitely an improvement. Before this change, it looks like we used `UTC` time. Do we want to present the date as local time in this case?

original problem: `SELECT CURRENT_TIMESTAMP` -> `YYYY-MM-DD HH:MM:SS` in utc but without zulu appendix. `ISO()` does not ensure zulu but `new Date()` then assumes it's local time -> we get a timestamp with the offset doubled

ms precision is additional, but prevents a fix in `ISO()` (cf. https://github.com/cap-js/cds-dbs/pull/254#issue-1914324847)

File: db-service/lib/SQLService.js (hunk 2 above)
Comment (johannes-vogel) [reply db-service/lib/SQLService.js]: I've double checked and the function `onPlainSQL` is overridden in postgres.

**Comment by johannes-vogel** [comment]
> > @danjoa @BobdenOs @patricebender any opinions?
> 
> If the requirement is to get to `ms` precision, this is definitely an improvement. Before this change, it looks like we used `UTC` time. Do we want to present the date as local time in this case?

We always assume that the timestamps are in zulu timezone. the current behavior is unexpected as the current_timestamp of a view definition looks like zulu time zone but is two hours off. We definetely must fix the time zone. ms precision is a nice side effect imo.

File: db-service/lib/SQLService.js (hunk 2 above)
Comment (sjvans) [reply db-service/lib/SQLService.js]: then we should probably check why deploy to postgres reached there...

**Review by johannes-vogel — APPROVED** [review]

**Pull request merged**